examples/maze-solver/maze.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2019 D-Wave Systems Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import dwavebinarycsp as dbc
+from re import match
+
+
+def build_bqm(n_rows, n_cols, start, end, walls, penalty_per_tile=0.5):
+    """Returns a BQM that corresponds to a valid path through a maze. This maze is described by the parameters.
+
+    Specifically, it uses the parameters to build a maze constraint satisfaction problem (CSP). This maze CSP is then
+    converted into the returned BQM.
+
+    Note: If penalty_per_tile is too large, the path will be too heavily penalized and the optimal solution might
+    produce no path at all.
+
+    Args:
+        n_rows: Integer. The number of rows in the maze.
+        n_cols: Integer. The number of cols in the maze.
+        start: String. The location of the starting point of the maze. String follows the format of get_label(..).
+        end: String. The location of the end point of the maze. String follows the format of get_label(..).
+        walls: List of Strings. The list of inner wall locations. Locations follow the format of get_label(..).
+        penalty_per_tile: A number. Penalty for each tile that is included in the path; encourages shorter paths.
+
+    Returns:
+        A dimod.BinaryQuadraticModel
+    """
+    maze = Maze(n_rows, n_cols, start, end, walls)
```

Comment:
Constructing a `Maze` object, just to discard it after constructing a BQM seems almost wasteful.

But I love the level of encapsulation you achieved it through that.